src/Microsoft.AspNetCore.Routing/Matching/DfaMatcherBuilder.cs

```diff
@@ -145,72 +147,80 @@ public DfaNode BuildDfaTree(bool includeLabel = false)
+                                // on the catchall node because only catchalls can end up here. The
+                                // catchall node allows us to capture an unlimited amount of segments
+                                // and also to match a zero-length segment, which a parameter node
+                                // doesn't allow.
+                                if (parent.CatchAll == null)
+                                {
+                                    parent.CatchAll = new DfaNode()
+                                    {
+                                        PathDepth = parent.PathDepth + 1,
+                                        Label = includeLabel ? parent.Label + "{*...}/" : null,
+                                    };
+
+                                    // The catchall node just loops.
+                                    parent.CatchAll.Parameters = parent.CatchAll;
+                                    parent.CatchAll.CatchAll = parent.CatchAll;
+                                }
+
+                                parent.CatchAll.AddMatch(endpoint);
                             }
-
-                            parent.CatchAll.AddMatch(endpoint);
-                        }
-                        else if (segment.IsSimple && part.IsParameter)
-                        {
-                            if (parent.Parameters == null)
+                            else if (RequiredValueHelpers.TryGetRequiredValue(endpoint.RoutePattern, parameterPart, out var requiredValue))
                             {
-                                parent.Parameters = new DfaNode()
+                                if (endpoint.RoutePattern.ParameterPolicies.TryGetValue(parameterPart.Name, out var parameterPolicyReferences))
                                 {
-                                    PathDepth = parent.PathDepth + 1,
-                                    Label = includeLabel ? parent.Label + "{...}/" : null,
-                                };
+                                    for (var k = 0; k < parameterPolicyReferences.Count; k++)
+                                    {
+                                        var reference = parameterPolicyReferences[k];
+                                        var parameterPolicy = _parameterPolicyFactory.Create(parameterPart, reference);
+                                        if (parameterPolicy is IOutboundParameterTransformer parameterTransformer)
+                                        {
+                                            requiredValue = parameterTransformer.TransformOutbound(requiredValue);
+                                            break;
+                                        }
+                                    }
+                                }
+
+                                AddLiteralNode(includeLabel, nextParents, parent, requiredValue.ToString());
```

Comment:
In a case like `Foo/{action=Index}` where `rv: { action = "Index" }`  how does `Foo` match? Is that handled by the handling for default values? This is worth a comment at least.

Reply:
> Is that handled by the handling for default values?

Yes, HasAdditionalRequiredSegments will add the endpoint as a match for the current node. I don't think this needs a comment to call that out here. There is no difference between how that logic interacts with required values and anything else.

I'll add a comment about the what is happening to the required value before it is used as a literal.